meeting scheduling algorithm and tests for calendar project

emaohn opened the pull request.

ganewman42 commented on walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java:
```diff
@@ -22,6 +22,13 @@
+    //   if (name.charAt(i) != ' ') {
+    //     result = name.substring(i);
+    //   }
+    // }
```
For cleanliness, better to delete this than comment it out. If you might want it later you can put it on another branch instead.

ganewman42 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+    Collection<String> MandatoryAttendants = request.getAttendees();
+    Collection<String> optionalAttendants = request.getOptionalAttendees();
+    for(Event e: events) {
+      boolean checkMandatoryAttendee = MandatoryAttendants.size() == 0 ? false : true;
```
This variable confuses me a bit. Isn't this redundant because if `MandatoryAttendants` is empty, it will never contain the attendee? Why do you override it lower down when it will just be changed by the next iteration?

emaohn replied:
The variable was intended to just ensure that I don't look for extra mandatory attendees who will be attending the event and perform the same splitting computation since we'd already know the event time causes conflict for the meeting. But yes I agree the initialization is pretty redundant so I'll remove that and also rename the variable to make it more clear! Thanks for catching it.

ganewman42 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+    return overlapTimeRanges;
   }
+
+    private TimeRange getOverlap(TimeRange mandatoryRange, TimeRange optionalRange) {
```
This is nicely laid out.

ganewman42 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+import java.util.*; 
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Nit: It took me a bit to figure out what events meant in this context, using a variable name like existingEvents or previousEvents might help make the code a little more clear.

emaohn replied:
That makes sense--this was just the method header already given to us in the starter code but I agree and I was confused as well when I started.

ganewman42 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+    ArrayList<TimeRange> overlapTimeRanges = new ArrayList<TimeRange>();
+    int mandatoryIndex = 0;
+    int optionalIndex = 0;
+    while (mandatoryIndex < mandatoryTimeRanges.size() && optionalIndex < optionalTimeRanges.size()) {
```
This loop seems to rely on the time ranges being sorted, but I don't see a sort or anything else that guarantees order in here (though could be I'm missing it). Make sure that your assumptions are guaranteed or you can get weird edge case behavior.

emaohn replied:
My algorithm for modifying the timeranges list should ensure that they're being split and inserted in an ordered manner. If you go down a bit you'll see the insertion part when I add to the timeranges. Maybe there is a bug that I don't know of?

cthamrun replied:
One good side note here is that when there are implicit/not immediately obvious implications (e.g. time ranges will be sorted), it would be good to add it as a comment to make it explicit to make things clearer.

ganewman42 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+                    if (eventRange.start() - curRange.start() >= request.getDuration()) {
+                        timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(curRange.start(), eventRange.start(), false));                    }
+                    if (curRange.end() - eventRange.end() >= request.getDuration()) {
+                    timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(eventRange.end(), curRange.end(), false));
```
Nit: Please indent

ganewman42 commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+    }
+
+    private void handleEventConflict(ArrayList<TimeRange> timeranges, Event e, MeetingRequest request) {
+        for (int i = 0; i < timeranges.size();) {
```
So what you're doing here is iterating over indices and deleting items. I believe this is going to be problematic because as you remove items, the list gets shorter and the indices mean different things since ArrayLists are dynamic.

As an example, I think this snippet:
```
ArrayList<String> x = new ArrayList<String>(Array.asList("hi", "bye"));
for (int i = 0; i < x.size();) {
  System.out.println(i);
  System.out.println(x[i]);
  x.remove(i);
  System.out.println(x[i]);
}
```

Would output:
```
0
hi
bye
```

Definitely double check this belief since Java is not my best language. Here's a discussion I used as a source that might be helpful if you need to do something else: https://stackoverflow.com/questions/10431981/remove-elements-from-collection-while-iterating

emaohn replied:
Yes you're correct! I believe I handled this situation however..which is why I don't automatically indent i (i++ in the for loop header) and instead manually increment i as necessary. Perhaps this isn't best code practice?

ganewman42 replied:
Yeah it might not be the best practice. Admittedly I was pretty tired when I read this CL, but I found it a bit confusing. That kind of deletion sends up a red flag to a reader because it's a common mistake. Even if you've mitigated it and everything works fine, you don't want to use patterns that read as a bug to a casual observer. Think about using a different deletion pattern.

emaohn replied:
Sounds good, just implemented an alternative solution.

cthamrun commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,5 +272,111 @@ public void notEnoughRoom() {
+  }
+
+  @Test
+  public void optionalAttendeesAreConsidered() {
```
Nit: consider adding a test scenario to reflect/test the behavior where there is a time slot that works for person A and person B but none that works for person C (I expect that we ignore person C since that person is optional and just schedule as per usual between person A and B).

emaohn replied:
Yes! I believe the test everyMandatoryAtendeesIsPrioritized() tests that case.

ganewman42 commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,5 +272,111 @@ public void notEnoughRoom() {
-}
 
+  @Test
+  public void everyMandatoryAttendeesIsPrioritized() {
```
Nice test coverage.

cthamrun commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+    optionalTimeRanges.add(TimeRange.WHOLE_DAY);
+    Collection<String> MandatoryAttendants = request.getAttendees();
+    Collection<String> optionalAttendants = request.getOptionalAttendees();
+    for(Event e: events) {
```
Nit: space after "for" and space after "e" (i.e. "for (Event e : events) {") for consistency within the file.

cthamrun commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+    for(Event e: events) {
+      boolean checkMandatoryAttendee = MandatoryAttendants.size() == 0 ? false : true;
+      int numOptionalUnavailable = 0;
+      for(String attendee: e.getAttendees()) {
```
Nit: same here as comment above.

cthamrun commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+                        timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(curRange.start(), eventRange.start(), false));                    }
+                    if (curRange.end() - eventRange.end() >= request.getDuration()) {
+                    timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(eventRange.end(), curRange.end(), false));
+                    }
```
Nit: make sure indentation and curly braces placement is consistent in this chunk here.

cthamrun commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+                } else if (curRange.contains(eventRange)) {
+                    int numNewRanges = 0;
+                    if (eventRange.start() - curRange.start() >= request.getDuration()) {
+                        timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(curRange.start(), eventRange.start(), false));                    }
```
Nit/optional: consider moving incrementation of numNewRanges to line above and separate it from add function call for slightly clearer readability.

cthamrun commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,118 @@
+                    }
+                    timeranges.remove(i);
+                    i += numNewRanges;
+                // case 3: slight slight overlap: either event starts before timerange or starts during and continues past timerange
```
"slight slight" typo?

emaohn commented:
I fixed the errors that you pointed out and added some commenting for clarification although perhaps my algorithm to split and re-insert times into the time ranges is messy and incomprehensible?

emaohn commented:
I changed handleEventConflict to return a new arraylist of timeranges that I construct as I look for conflicts and resolve them by adding working ranges into the result rather than deleting and inserting into the original timerange list for better readability.
